### PR TITLE
Allow creating runs

### PR DESCRIPTION
**Pull request by pbugnion**
This PR allows creating runs with `mlflow.start_run`.

To test this on a faculty server
- uninstall faculty to force a reinstallation
- install mlflow-faculty on this branch -- this should pull faculty master and mlflow master.

```py
import os
os.environ["MLFLOW_TRACKING_URI"] = "faculty:{}".format(os.environ["FACULTY_PROJECT_ID"])
import mlflow

mlflow.start_run()
```

Tested this against currently deployed atlas and it works.

### CODE REVIEW

File: tests/test_trackingstore.py [hunk 1]
```diff
@@ -191,3 +192,61 @@ def test_list_experiments_client_error(mocker):
...
     with pytest.raises(MlflowException, match="Error"):
         store.list_experiments(PROJECT_ID)
+
+
+def test_create_run(mocker):
+    mock_client = mocker.Mock()
+    mock_client.create_run.return_value = None
+    mocker.patch("faculty.client", return_value=mock_client)
+    mock_mlflow_run = mocker.Mock()
+    mocker.patch(
```
Comment (tomas-milata) [inline tests/test_trackingstore.py]: If we're mocking out `faculty_run_to_mlflow_run`, should we also check that `faculty_run_to_mlflow_run` was called with the right argument, i.e. with the value returned from the client?

Comment (pbugnion) [reply tests/test_trackingstore.py]: good call.

**Comment by pbugnion** [comment]
This is ready for review now.

File: mlflow_faculty/mlflow_converters.py [hunk 2]
```diff
@@ -1,4 +1,29 @@
...
+)
+
+from faculty.clients.experiment import (
+    ExperimentRunStatus as FacultyExperimentRunStatus,
+)
+
+from mlflow_faculty.py23 import to_timestamp
+
+
+_run_status_map = {
```
Comment (acroz) [inline mlflow_faculty/mlflow_converters.py]: I think Python style conventions would have this named `_RUN_STATUS_MAP`?

File: mlflow_faculty/mlflow_converters.py [hunk 3]
```diff
@@ -1,4 +1,29 @@
-from mlflow.entities import Experiment, LifecycleStage
+from mlflow.entities import (
+    Experiment,
+    LifecycleStage,
+    RunInfo,
+    RunStatus,
+    Run,
+    RunData,
+)
+
```
Comment (acroz) [inline mlflow_faculty/mlflow_converters.py]: Remove this line? (i.e. group 'external' library imports)

File: mlflow_faculty/mlflow_converters.py [hunk 4]
```diff
@@ -9,3 +34,34 @@ def faculty_experiment_to_mlflow_experiment(faculty_experiment):
...
+        faculty_run.experiment_id,
+        "",  # name
+        "",  # source_type
+        "",  # source_name
+        "",  # entry_point_name
+        "",  # user_id
+        _run_status_map[faculty_run.status],
+        start_time,
+        end_time,
+        "",  # shource version
```
Comment (acroz) [inline mlflow_faculty/mlflow_converters.py]: shource -> source

File: mlflow_faculty/py23.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import six
+from pytz import UTC
+from datetime import datetime
+
+EPOCH = datetime(1970, 1, 1, tzinfo=UTC)
+
```
Comment (acroz) [inline mlflow_faculty/py23.py]: Add extra empty line (I believe this would currently fail `flake8`?)

File: mlflow_faculty/py23.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+def to_timestamp(dt):
+    if six.PY2:
+        return (dt - EPOCH).total_seconds()
+    else:
+        return dt.timestamp()
+
```
Comment (acroz) [inline mlflow_faculty/py23.py]: Remove empty line at EOF.

File: mlflow_faculty/trackingstore.py [hunk 7]
```diff
@@ -13,6 +13,8 @@
 # limitations under the License.
 
 from uuid import UUID
+from datetime import datetime
+from pytz import UTC
```
Comment (acroz) [inline mlflow_faculty/trackingstore.py]: Separate `pytz` import from standard library imports?

File: mlflow_faculty/trackingstore.py [hunk 8]
```diff
@@ -199,7 +201,21 @@ def create_run(
 
         :return: The created Run object
         """
-        raise NotImplementedError()
+        try:
+            faculty_run = self._client.create_run(
+                self._project_id,
+                experiment_id,
+                datetime.fromtimestamp(start_time / 1000, tz=UTC),
```
Comment (acroz) [inline mlflow_faculty/trackingstore.py]: Did you consider putting this logic as a helper in the converters for clear symmetry with the opposite conversion logic?

File: tests/test_mlflow_converters.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    started_at=RUN_STARTED_AT,
+    ended_at=None,
+    deleted_at=None,
+)
+
+
+EXPECTED_RUN_INFO = RunInfo(
+    EXPERIMENT_RUN_ID,
+    661,
+    "",  # name
```
Comment (tomas-milata) [inline tests/test_mlflow_converters.py]: I know it's probably not relevant to this PR, but why does the faculty -> mlflow converter have a lot of these empty strings? Do we not have access to values for these fields at all?  Is the plan to have e.g. `user_id` available in future?

File: mlflow_faculty/trackingstore.py [hunk 10]
```diff
@@ -199,7 +201,21 @@ def create_run(
...
+            )
+        except faculty.clients.base.HttpError as e:
+            raise MlflowException(
+                "{}. Received response {} with status code {}".format(
```
Comment (acroz) [inline mlflow_faculty/trackingstore.py]: Should we have something more specific e.g. 'Failed to create run.' as part of the error message?

File: tests/test_mlflow_converters.py (hunk 9 above)
Comment (pbugnion) [reply tests/test_mlflow_converters.py]: Good question. These values are stored in tags, which are currently not returned by the backend.

In milestone 1.3, we'll add tags to resolve these.

E.g. `entry_point = run.tags.get("mlflow.project.entryPoint")`

The tags we'll add will include:

```
"mlflow.runName" -> string
"mlflow.parentRunId" -> UUID | null
"mlflow.source.type" -> "local | notebook | job | project | unknown"
"mlflow.source.name" -> "path-to-notebook"
"mlflow.source.version" -> "git-hash" | null
"mlflow.gitRepoURL" -> string | null
"mlflow.gitBranchName" -> string | null
"mlflow.project.entryPoint" -> "name-of-entry-point" | null
"mlflow.faculty.created.by" -> "job" | "user"
"mlflow.faculty.created.user.userId" -> UUID | null
"mlflow.faculty.created.job.jobId" -> UUID | null
"mlflow.faculty.created.job.runId" -> UUID | null
"mlflow.faculty.created.job.subrunId" -> UUID | null
"mlflow.user" -> string | null
```

Comment (tomas-milata) [reply tests/test_mlflow_converters.py]: I see, thanks!

File: mlflow_faculty/trackingstore.py (hunk 10 above)
Comment (tomas-milata) [reply mlflow_faculty/trackingstore.py]: How about creating a custom exception class, e.g. `FacultyStoreException` that takes 2 constructor arguments
 - message (e.g. 'Failed to create run')
 - a `faculty.clients.base.HttpError`

and extends `MlflowException`

and does the `"{}. Received response {} with status code {}".format()...` inside ?

There are other store-specific exceptions in the mlflow codebase, such as `RestException`.

Comment (acroz) [reply mlflow_faculty/trackingstore.py]: I'm not sure if this is super valuable beyond just having the error message inside a standard MlflowException. I would advocate using MlflowException for now and subclassing later if it is considered valuable.

File: mlflow_faculty/trackingstore.py (hunk 8 above)
Comment (pbugnion) [reply mlflow_faculty/trackingstore.py]: done

File: mlflow_faculty/trackingstore.py (hunk 10 above)
Comment (tomas-milata) [reply mlflow_faculty/trackingstore.py]: We could have helper method instead, that just instantiates an MlflowException, but I gues one of my points was to do the `"{}. Received response {} with status code {}"` on one common place and the using more descriptive error messages on the specific places.

**Comment by pbugnion** [comment]
> Should there also be test(s) for the `py23` module?

Yes -- done

File: mlflow_faculty/trackingstore.py (hunk 10 above)
Comment (pbugnion) [reply mlflow_faculty/trackingstore.py]: I'm going to construct the exception manually for now. If we re-use similar code in different places, we can refactor it later.

**Pull request merged**